### PR TITLE
[WIP] Ssl stream bug hunt - trying to find the outerloop failure

### PR DESCRIPTION
**Pull request by Drawaes**
I have just free'd the GCHandle rather than setting it to null.

### CODE REVIEW

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
DONT REVIEW YET... This is just a full outerloop test.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs [hunk 1]
```diff
@@ -389,9 +389,11 @@ private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out byte* outp, o
             }
             catch
             {
+                protocolHandle.Free();
                 return Ssl.SSL_TLSEXT_ERR_NOACK;
```
Comment (Priya91) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: This doesn't work, it actually doesn't free the original handle, it just frees the copy, you should unskip the nonmatching alpn protocol test to verify it passes with the change.

**Comment by Priya91** [comment]
Btw, I was able to repro the timeout under debugger on windows, what's happening is the WriteAsync doesn't hang, but i timeout on the readasync virtualnetwork timeout, so readasync times out, even before writeasync finishes operation. It could be the case that sometimes the readasync timeout value is low. The writeasync technically times out at testpassingtimeoutmilliseconds, which is around 4*60 sec, but the readasync virtualnetwork timeout is around 30s. So if the test fails we want it to fail on writeasync timeout, rather than the serverstream readasync timing out. I'm going to add some log statements to verify if this is what is happening, if so, then change the test to timeout on writeasync instead of readasync.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs (hunk 1 above)
Comment (Drawaes) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: Its not how GCHandle works. While it's true the struct is copied, it is actually a slot (or in the code a cookie) in the GC with a ref to the object.

If you "free" a copy, it doesn't update the "allocated" state on the original struct, but it does actually mean the original will see "null" as it's target. Check the below example code

``` csharp
static void Main(string[] args)
{
        var handle = GCHandle.Alloc(new byte[100]);

        var ptr = GCHandle.ToIntPtr(handle);

        var handle2 = GCHandle.FromIntPtr(ptr);
        Console.WriteLine(handle.Target == null);
        handle2.Free();

        Console.WriteLine($"handle target is null {handle.Target == null}");
}
```

This shows that freeing the copy actually makes the original's "target" property == null.

Comment (Drawaes) [reply src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: @stephentoub actually brought up the fact that I am not 100% sure on, that you can change the target and that works, in theory it should but I have never seen anyone "reallocate" the pinned object on the slot after it has been allocated. If it doesn't work, then it's potentially a bug, or a mistake to be allowed to "set" the target property.

**Comment by Drawaes** [comment]
FYI I have no intention of this being merged, its just an experiment to see if it passes the outerloop, if it doesn't then we can try and enable the tests and see if it does or doesn't. I am basically performing a binary search at this point.

**Comment by Drawaes** [comment]
The outerloop passed (well Redhat never actually started but all other distros passed) so I am kicking it off with the tests (they might not pass due to the win 7/distros that don't support alpn) but I am looking for the Async/Sync outerloop failures not for the ALPN results

@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
Closed as @Priya91 fixed the test in #25261

**Comment by Drawaes** [comment]
Checking if Init Crypto helps here

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
Init made no diff

**Comment by Drawaes** [comment]
Bringing it back, funnily I had a similar issue to this (no in the same code) with testing at work, on Linux and its just clicked. So I am going to try some ideas out.

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
No intention to close this, it did it automagically when I did a reset --hard from master to reset my testing.

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
@dotnet-bot Test outerloop Linux x64 Debug Build please

**Comment by Drawaes** [comment]
/cc @stephentoub @Priya91  

I am just leaving the code from the ALPN fix in here, so that you can see the results.

**Comment by Drawaes** [comment]
Closing this as it was just to experiment as to where the issue with failing outerloop was

**Pull request closed without merging**